### PR TITLE
Refactor int8 dynamic quantization with call to `quantize`

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
Previously we added `quantize` as a general API (https://github.com/pytorch/ao/pull/256) for Affine Quantized tensor subclass, and also tensor subclass based dtype conversion in general.

The plan is to use this to replace existing quant APIs including int4 weight only, int8 weight only, int8 dynamic quant and 8da4w (for executorch).

This PR we started replacing the implementation of int8 dynamic quant API with `quantize` API with affine quantized tensor subclass. We'll make sure the performance does not regress for vit model.

Next:
* refactor int4 weight only and int8 weight only apis
* add some unit tests for affine quantized tensor and layout tensor subclasses
* 8da4w can happen a bit later since we need to coordinate with executorch team on how to make the switch

Test Plan:
TORCH_LOGS='output_code' python tutorials/quantize_vit/run_vit_b_quant.py

reference: elapsed_time:  1.4821058654785155  milliseconds
after refactor: elapsed_time:  1.4826457214355468  milliseconds

generated code diff: https://gist.github.com/jerryzh168/edca9d421363582b66d41c9cc2db0f7b
(meta only) paste diff: https://www.internalfb.com/phabricator/paste/view/P1385188208

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torchao/quantization/quant_api.py [hunk 1]
```diff
@@ -187,9 +191,13 @@ def change_linear_weights_to_int8_dqtensors(model, filter_fn=None, **kwargs):
             *args
         )
 
-    _replace_with_custom_fn_if_matches_filter(
-        model, _get_subclass_inserter(Int8DynamicallyQuantizedLinearWeight, enable_parametrization=TORCH_VERSION_AFTER_2_4, **kwargs), filter_fn
-    )
+    if TORCH_VERSION_AFTER_2_4:
```
Comment (cpuhrsch) [inline torchao/quantization/quant_api.py:194]: Why is this needed?

File: torchao/dtypes/aqt.py [hunk 2]
```diff
@@ -282,6 +288,74 @@ def get_plain(self):
             f"Unpacking for tensor core tiled storage is not yet implemented"
         )
 
+@register_aqt_layout_cls("transposed")
+class TransposedAQTLayout(PlainAQTLayout):
```
Comment (cpuhrsch) [inline torchao/dtypes/aqt.py:292]: Why not use strides?

Comment (jerryzh168) [reply torchao/dtypes/aqt.py:292]: main thing is that we need to transpose the weight before hand, while logically the weight still has the same dimensions, I think it's just a kind of packing. what do you mean by using strides? (doesn't transpose also just change the strides?)

Comment (cpuhrsch) [reply torchao/dtypes/aqt.py:292]: Yes, don't transpose and just use the strides

Comment (cpuhrsch) [reply torchao/dtypes/aqt.py:292]: Or put differently, when you implement `transpose` you change the sizes and strides so that it can be a view operation. It makes sense that perhaps for the packed layout you can't have strides, but this seems to be just the usual triplet?

Comment (jerryzh168) [reply torchao/dtypes/aqt.py]: OK, since `transpose` is cheap, we'll just do this when we run the kernel, removed the transposed layout

File: test/quantization/test_quant_api.py [hunk 3]
```diff
@@ -493,7 +513,7 @@ def test_quantized_tensor_subclass_int8(self):
     @unittest.skipIf(not torch.cuda.is_available(), "Need CUDA available")
     def test_quantized_tensor_subclass_int8_dyn_quant(self):
         # use 1024 so that we don't need padding
```
Comment (msaroufim) [inline test/quantization/test_quant_api.py:515]: nit: comment needs updating, how come this was changed?

Comment (jerryzh168) [reply test/quantization/test_quant_api.py:515]: oh I was testing around transpose related things before, I can update the comment, it's good to have different out/in dimensions for weight to make sure we also test the transpose I think

File: test/quantization/test_quant_api.py [hunk 4]
```diff
@@ -525,6 +545,48 @@ def test_quantized_tensor_subclass_int8_dyn_quant(self):
...
+        m = torch.compile(m, mode='max-autotune', fullgraph=True)
+
+        benchmark_model(m, WARMUP, input_tensor)
+        elapsed_time = benchmark_model(m, RUNS, input_tensor)
+
+        m_ref = torch.compile(m_ref, mode='max-autotune', fullgraph=True)
+        benchmark_model(m_ref, WARMUP, input_tensor)
+        ref_elapsed_time = benchmark_model(m_ref, RUNS, input_tensor)
+
+        # recent measurement result:
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/quantization/test_quant_api.py:582]: yeah this is in A100, I can remove these as well

File: test/quantization/test_quant_api.py [hunk 5]
```diff
@@ -525,6 +545,47 @@ def test_quantized_tensor_subclass_int8_dyn_quant(self):
...
+        # elapsed time: 0.2358572769165039, ref elapsed time: 0.2810140800476074
+        # elapsed time: 0.22115007400512696, ref elapsed time: 0.32116992950439455
+        # elapsed time: 0.2349068832397461, ref elapsed time: 0.2812371253967285
+        print(f"elapsed time: {elapsed_time}, ref elapsed time: {ref_elapsed_time}")
+        self.assertTrue(elapsed_time < 1.05 * ref_elapsed_time)
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/quantization/test_quant_api.py:587]: this is pretty quick when I run it in my A100 machine, finishes in a few seconds. I could also skip this by default and just have people run this locally when making changes to these APIs

File: torchao/dtypes/aqt.py [hunk 6]
```diff
@@ -576,45 +577,77 @@ def functional_linear(*args, **kwargs):
...
-        )
-        input_tensor, weight_qtensor, bias = (
+        input_tensor, weight_tensor, bias = (
             args[1],
             args[2],
             args[0],
         )
+        try:
+            return _quantized_linear_op(input_tensor, weight_tensor, bias)
+        except:
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torchao/dtypes/aqt.py:636]: oh we actually need to call the function in different ways here, not sure when the change is reverted, will fix

try except is used as a fallback when the specific configuration of input and weight tensor is not caught by any of the special dispatches in `_quantized_linear_op`

File: torchao/quantization/subclass.py [hunk 7]
```diff
@@ -704,6 +730,19 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
...
+        if func is aten._to_copy.default:
+            return return_and_correct_aliasing(
+                func,
+                args,
+                kwargs,
+                args[0].to(*args[1:], **kwargs)._apply_fn_to_data(torch.clone),
+            )
+
+        if func is aten.t.default:
+            return return_and_correct_aliasing(
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torchao/quantization/subclass.py:742]: the code is mostly copy pasted from other places, but looks like it's a trick for tensor subclass to work with torch.compile and make sure we have correct aliasing behavior: https://github.com/pytorch/pytorch/blob/214dd44608a92802f9c13471451ae09cf6b25fd0/torch/utils/_python_dispatch.py#L518

File: torchao/dtypes/aqt.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torchao/dtypes/aqt.py:642]: added some comments

File: torchao/dtypes/aqt.py [hunk 8]
```diff
@@ -576,45 +582,83 @@ def functional_linear(*args, **kwargs):
...
-    weight_tensor = weight_qtensor.dequantize()
-    return func(input_tensor, weight_tensor, bias)
+        try:
+            return _quantized_linear_op(input_tensor, weight_tensor, bias)
+        except:
+            if isinstance(input_tensor, AffineQuantizedTensor):
+                input_tensor = input_tensor.dequantize()
+            if isinstance(weight_tensor, AffineQuantizedTensor):
+                weight_tensor = weight_tensor.dequantize()
+            return func(input_tensor, weight_tensor)
```
Comment (jerryzh168) [inline torchao/dtypes/aqt.py:661]: so here is a difference of how we call the function, since we have aten.mm here, the order of passing around the args are different from aten.addmm

File: test/quantization/test_quant_api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/quantization/test_quant_api.py:584]: skipped this one by default

**Review by msaroufim — APPROVED** [review]

File: test/quantization/test_quant_api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cpuhrsch) [reply test/quantization/test_quant_api.py:584]: Yeah doing benchmarks in unit tests is a known anti pattern. Test environments don't need to be inconsistent and it's likely a waste of resources to make them so.
